grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1132,106 +1140,438 @@ class StockMovementService {
+            }
+            // Scenario 3b: if lot has stock entries in default bin
+            // TODO If there's enough quantity available in default location, then
+            //  allocate from default location.
+
+            // Scenario 3c: if lot has stock entries only in hold bins
+            // TODO If lot has stock entries only in hold bin, then throw an exception.
+
+            // Scenario 4: Any other scenario (stock in “real” bin and virtual bin, stock in multiple real bins)
+            // TODO If there's not enough available quantity in default + receiving
+            //  locations, then throw an exception
+            boolean isAllStockInReceivingOrDefault = availableItems.every { it.isReceivingLocation || it.isDefaultLocation }
+            if (isAllStockInReceivingOrDefault) {
+                return availableItems
+            }
+        }
+        return []
+    }
+
+    void importPicklistItems(PicklistImportDataCommand command) {
+
+        StockMovement stockMovement = command.stockMovement
+        List<PickPageItem> pickPageItems = command.pickPageItems
+        List<PicklistItemCommand> picklistItems = command.picklistItems
+
+        Map<String, List<PicklistItemCommand>> picklistItemsGroupedByRequisitionItem = picklistItems.groupBy { it.id }
 
-        picklistItemEntriesGroupedByRequisitionItem.each { requisitionItemId, picklistsToImport  ->
+        picklistItemsGroupedByRequisitionItem.each { String requisitionItemId, List<PicklistItemCommand> picklistItemsToImport  ->
             // skip rows with errors
-            if (picklistsToImport.any{ it.hasErrors() }) {
+            if (picklistItemsToImport.any{ it.hasErrors() }) {
                 return
             }
 
             PickPageItem pickPageItem = pickPageItems.find {
                 it.requisitionItem?.id == requisitionItemId
             }
 
+            // TODO Does the order of operations matter (remove shipment items vs picklist items)
+            // Remove existing shipment items and picklist items
             removeShipmentItemsForModifiedRequisitionItem(pickPageItem.requisitionItem)
 
-            picklistsToImport.each { params ->
-                AvailableItem availableItem = pickPageItem.getAvailableItem(params.binLocation, params.lotNumber)
-                Picklist picklist = stockMovement.requisition?.picklist
-                // find existing picklist to update
-                PicklistItem picklistItem = picklist.picklistItems.find {
-                    it.inventoryItem?.id == availableItem.inventoryItem?.id &&
-                            it.binLocation?.id == availableItem.binLocation?.id
+            // TODO Need to test this thoroughly as it does not feel like a good idea. For example,
+            //  we don't necessarily want to clear picklist if there's no picklist items to import
+            //  so this might need to go inside a check on !picklistItemsToImport.empty
+            clearPicklist(pickPageItem.requisitionItem)
```

Comment:
Yeah I think this might cause unnecessary product availability refresh if there are no existing picklist items at this stage